### PR TITLE
nsolid: add support for start/stopTrackingHeapObjects

### PR DESCRIPTION
**Pull request by juanarbol**
This patch implements triggers for `startTrackingHeapObjects` and `stopTrackingHeapObjects` in the heap profiler.

The implementation is based on what the inspector does:

The `startTrackingHeapObjects` will simply activate that in the HeapProfiler and; the `stopTrackingHeapObjects` will create a heap snapshot and deactivate the `TrackingHeapObjects` in the Heap profiler.

This patch also includes the `trackAllocations` option supported.

### CODE REVIEW

File: src/nsolid.h [hunk 1]
```diff
@@ -961,7 +962,45 @@ class NODE_EXTERN Snapshot {
                           Cb&& cb,
                           Data&&... data);
 
+  /**
+   * @brief activates the heap profiler tracking objects from a specific JS
+   * thread.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param redacted to get redacted strings in the heap snapshot.
+   * @param data variable number of arguments to be propagated to the callback.
```
Comment (santigimeno) [inline src/nsolid.h:985]: Which cb are we propagating to?

File: src/nsolid.h [hunk 2]
```diff
@@ -961,7 +962,45 @@ class NODE_EXTERN Snapshot {
...
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
+   */
+  template <typename... Data>
+  static int StartTrackingHeapObjects(SharedEnvInst envinst,
+                                      bool trackAllocations,
+                                      Data&&... data);
+
```
Comment (santigimeno) [inline src/nsolid.h]: I'd change the interface a bit and make it more like the `CpuProfile`:
```suggestion
  template <typename Cb, typename... Data>
  static int TrackHeapObjects(SharedEnvInst envinst,
                              bool trackAllocations,
                              uint64_t duration,
                              Cb&& cb,
                              Data&&... data);

```
Then have a
```c++
static int StopTrackingHeapObjects(SharedEnvInst envinst);
```

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 3]
```diff
@@ -38,6 +68,16 @@ int NSolidHeapSnapshot::GetHeapSnapshot(SharedEnvInst envinst,
...
 }
 
+void NSolidHeapSnapshot::start_tracking_heapobjects(
+    SharedEnvInst envinst,
+    bool trackAllocations,
+    NSolidHeapSnapshot* snapshotter) {
+  ASSERT_EQ(snapshotter->is_tracking_heapobjects_, false);
+  envinst->isolate()->GetHeapProfiler()->StartTrackingHeapObjects(
+      trackAllocations);
+  snapshotter->is_tracking_heapobjects_ = true;
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Check the return value just in case the inspector is running a profiler itself

Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: The `StartTrackingHeapObjects` is void, it does not returns anything

File: src/nsolid.h (hunk 1 above)
Comment (juanarbol) [reply src/nsolid.h:985]: Fixed in the latest revision! With the timer support!

File: src/nsolid.h (hunk 2 above)
Comment (juanarbol) [reply src/nsolid.h]: Fixed with the timers support

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 3 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Double-checked with Node.js and V8; nothing bad is going to happen.

File: src/nsolid.h [hunk 4]
```diff
@@ -961,7 +962,56 @@ class NODE_EXTERN Snapshot {
...
 
+  /**
+   * @brief activates the heap profiler tracking objects from a specific JS
+   * thread.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param redacted to get redacted strings in the heap snapshot.
+   * @param data variable number of arguments to be propagated to the callback.
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
```
Comment (santigimeno) [inline src/nsolid.h]: It seems the arguments documentation is not updated

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 5]
```diff
@@ -31,13 +86,40 @@ int NSolidHeapSnapshot::GetHeapSnapshot(SharedEnvInst envinst,
...
+
+  envinst->isolate()->GetHeapProfiler()->StartTrackingHeapObjects(
+      trackAllocations);
+  if (duration > 0) {
+    int er = QueueCallback(duration, take_snapshot, envinst, snapshotter);
+
+    if (er) {
+      // In case the the thread is already gone, the cpu profile will be stopped
+      // on RemoveEnv, so do nothing here.
+    }
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc:176]: Is this still true for heap profiles?

File: src/nsolid.cc [hunk 6]
```diff
@@ -483,6 +483,28 @@ int CpuProfiler::get_cpu_profile_(SharedEnvInst envinst,
...
 }
 
+int Snapshot::start_tracking_heap_objects_(SharedEnvInst envinst,
+                                           bool redacted,
+                                           bool trackAllocations,
+                                           uint64_t duration,
+                                           void* data,
+                                           snapshot_proxy_sig proxy,
+                                           internal::deleter_sig deleter) {
+  if (envinst == nullptr) return UV_ESRCH;
```
Comment (santigimeno) [inline src/nsolid.cc]: Follow the syntax used in the rest of the code
```suggestion
  if (envinst == nullptr)
    return UV_ESRCH;
```

File: src/nsolid.cc [hunk 7]
```diff
@@ -483,6 +483,28 @@ int CpuProfiler::get_cpu_profile_(SharedEnvInst envinst,
...
+
+  return NSolidHeapSnapshot::Inst()->StartTrackingHeapObjects(
+      envinst, redacted, trackAllocations, duration, data, proxy, deleter);
+}
+
+int Snapshot::stop_tracking_heap_objects_(SharedEnvInst envinst,
+                                          void* data,
+                                          snapshot_proxy_sig proxy,
+                                          internal::deleter_sig deleter) {
+  if (envinst == nullptr) return UV_ESRCH;
```
Comment (santigimeno) [inline src/nsolid.cc]: Follow the syntax used in the rest of the code
```suggestion
  if (envinst == nullptr)
    return UV_ESRCH;
```

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 8]
```diff
@@ -12,6 +14,59 @@ NSolidHeapSnapshot::NSolidHeapSnapshot() {
...
+  // Consider this as taking a heap snapshot in the thread
+  if (status == 0) {
+    // Now we are tracking heap objects in this thread
+    threads_running_snapshots_.emplace(
+        thread_id,
+        HeapSnapshotStor{
+            redacted, true, proxy, internal::user_data(data, deleter)});
+  }
+
+  return status;
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: We can do a bit better:

```suggestion
  auto it = threads_running_snapshots_.emplace(
        thread_id,
        HeapSnapshotStor{
            redacted, true, proxy, internal::user_data(data, deleter)});

  if (it.second == false) {
    return UV_EEXIST;
  }

  int status = RunCommand(envinst,
                          CommandType::Interrupt,
                          start_tracking_heapobjects,
                          trackAllocations,
                          duration,
                          this);

  // Consider this as taking a heap snapshot in the thread
  if (status == 0) {
    // Now we are tracking heap objects in this thread
    threads_running_snapshots_.erase(it.first);
  }

  return status;
```

File: src/nsolid.h [hunk 9]
```diff
@@ -961,7 +962,56 @@ class NODE_EXTERN Snapshot {
...
+   * `cb(int status, std::string snapshot, ...Data)`. It will be called from the
+   * NSolid thread.
+   * @param data variable number of arguments to be propagated to the callback.
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
+   */
+  template <typename Cb, typename... Data>
+  static int StopTrackingHeapObjects(SharedEnvInst envinst,
+                                     Cb&& cb,
+                                     Data&&... data);
```
Comment (santigimeno) [inline src/nsolid.h]: As we discussed, no reason to have a cb for this as the completion will be notified using the `StartTrackingHeapObjects` cb.
```suggestion
  template <typename Cb, typename... Data>
  static int StopTrackingHeapObjects(SharedEnvInst envinst);
```

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 10]
```diff
@@ -12,6 +14,59 @@ NSolidHeapSnapshot::NSolidHeapSnapshot() {
...
+    bool trackAllocations,
+    uint64_t duration,
+    void* data,
+    Snapshot::snapshot_proxy_sig proxy,
+    internal::deleter_sig deleter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&in_progress_heap_snapshots_);
+  // We can not trigger this command if there is already a snapshot in progress
+  auto it = threads_running_snapshots_.find(thread_id);
+  if (it != threads_running_snapshots_.end()) {
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Are we using the same map because we're not allowing taking a heap profile while taking a snapshot and viceversa. If that's the case, we should document it, otherwise we should be using different map/lock

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: src/nsolid.cc (hunk 6 above)
Comment (juanarbol) [reply src/nsolid.cc]: Just letting you know, that is formatted by `make cpp-format`

File: src/nsolid.cc (hunk 7 above)
Comment (juanarbol) [reply src/nsolid.cc]: Just letting you know, that is formatted by `make cpp-format`

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 8 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Thanks! Will fix that!

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 10 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: When you start the tracking heap objects, we consider that as a “running/in progress” heap snapshot, that’s why I include that in the map/lock.

File: src/nsolid.h (hunk 9 above)
Comment (juanarbol) [reply src/nsolid.h]: Maybe?
```suggestion
  static int StopTrackingHeapObjects(SharedEnvInst envinst); // No template required here
```

File: src/nsolid.h (hunk 4 above)
Comment (juanarbol) [reply src/nsolid.h]: True that, on it!

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 5 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc:176]: NOTE: this does not makes sense, this is `QueueCallback` which is N|Solid thread, not JS thread.

File: src/nsolid.h [hunk 11]
```diff
@@ -961,7 +962,43 @@ class NODE_EXTERN Snapshot {
...
+   * thread.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param redacted to get redacted strings in the heap snapshot.
+   * @param data variable number of arguments to be propagated to the callback.
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
+   */
+  template <typename Cb, typename... Data>
+  static int StartTrackingHeapObjects(SharedEnvInst envinst,
```
Comment (santigimeno) [inline src/nsolid.h]: Still there params are not correctly documented

File: src/nsolid.h [hunk 12]
```diff
@@ -961,7 +962,43 @@ class NODE_EXTERN Snapshot {
...
+                                      bool redacted,
+                                      bool trackAllocations,
+                                      uint64_t duration,
+                                      Cb&& cb,
+                                      Data&&... data);
+
+  /**
+   * @brief same as TakeSnapshot but stops tracking heap objects in the heap
+   * profiler.
```
Comment (santigimeno) [inline src/nsolid.h:999]: It seems `TakeSnapshot` doesn't apply here.

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 8 above)
Comment (santigimeno) [reply src/nsolid/nsolid_heap_snapshot.cc]: This hasn't been resolved afaict

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 13]
```diff
@@ -31,13 +68,140 @@ int NSolidHeapSnapshot::GetHeapSnapshot(SharedEnvInst envinst,
 
   if (status == 0) {
     threads_running_snapshots_.emplace(
-      thread_id,
-      HeapSnapshotStor{ redacted, proxy, internal::user_data(data, deleter) });
+        thread_id,
+        HeapSnapshotStor{
+            redacted, false, proxy, internal::user_data(data, deleter)});
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc:65]: this change is unrelated to the current pr

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 14]
```diff
@@ -31,13 +68,140 @@ int NSolidHeapSnapshot::GetHeapSnapshot(SharedEnvInst envinst,
...
+  return er;
+}
+
+void NSolidHeapSnapshot::stop_tracking_heap_objects(
+    SharedEnvInst envinst,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_snapshots_);
+  auto it = snapshotter->threads_running_snapshots_.find(thread_id);
+  HeapSnapshotStor& stor = it->second;
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Still we should be checking whether there's a running snapshot here.

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: test/addons/nsolid-track-heap-objects/binding.cc [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+      thread_id);
+  args.GetReturnValue().Set(Integer::New(args.GetIsolate(), ret));
+}
+
+static void StopTrackingHeapObjects(const FunctionCallbackInfo<Value>& args) {
+  v8::HandleScope handle_scope(args.GetIsolate());
+  // thread_id
+  assert(args[0]->IsUint32());
+
+  uint64_t thread_id = args[0].As<Uint32>()->Value();
```
Comment (santigimeno) [inline test/addons/nsolid-track-heap-objects/binding.cc]: Being thread_id a `uint64_t` better be safe than sorry
```suggestion
  uint64_t thread_id = args[0].As<Number>()->Value();
```

File: test/addons/nsolid-track-heap-objects/binding.cc [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  // thread_id
+  assert(args[0]->IsUint32());
+  // Redacted heap snapshot
+  assert(args[1]->IsBoolean());
+  // Track allocations
+  assert(args[2]->IsBoolean());
+  // Stop after this many milliseconds
+  assert(args[3]->IsNumber());
+
+  uint64_t thread_id = args[0].As<Uint32>()->Value();
```
Comment (santigimeno) [inline test/addons/nsolid-track-heap-objects/binding.cc]: Same here, use `Number` is strictly more correct

**Review by santigimeno — CHANGES_REQUESTED** [review]
Looking very good

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 8 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: This suggestion provokes the runtime to abort. I'll proceed to resolve this conversation.

```lldb
(lldb) bt all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x000000018fee0180 libsystem_kernel.dylib`__kill + 8
    frame #1: 0x00000001015c78e8 nsolid`uv_kill(pid=0, signum=6) at process.c:1091:7
    frame #2: 0x00000001002edc78 nsolid`node::Kill(args=0x000000016fdfcfe8) at node_process_methods.cc:166:13
    frame #3: 0x0000000100788d8c nsolid`v8::internal::FunctionCallbackArguments::Call(this=0x000000016fdfd088, handler=<unavailable>) at api-arguments-inl.h:146:3 [opt]
    frame #4: 0x0000000100787af8 nsolid`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(isolate=0x0000000110008000, new_target=Handle<v8::internal::HeapObject> @ x23, fun_data=<unavailable>, receiver=<unavailable>, argv=0x000000016fdfd228, argc=2) at builtins-api.cc:113:36 [opt]
    frame #5: 0x00000001007866e4 nsolid`v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*) [inlined] v8::internal::Builtin_Impl_HandleApiCall(args=<unavailable>, isolate=0x0000000110008000) at builtins-api.cc:144:5 [opt]
    frame #6: 0x00000001007865fc nsolid`v8::internal::Builtin_HandleApiCall(args_length=<unavailable>, args_object=<unavailable>, isolate=0x0000000110008000) at builtins-api.cc:135:1 [opt]
    frame #7: 0x0000000101b1e4a4 nsolid`Builtins_CEntry_Return1_ArgvOnStack_BuiltinExit + 100
    frame #8: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #9: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #10: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #11: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #12: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #13: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #14: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #15: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #16: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #17: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #18: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #19: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #20: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #21: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #22: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #23: 0x0000000101a80dec nsolid`Builtins_JSEntryTrampoline + 172
    frame #24: 0x0000000101a80ad4 nsolid`Builtins_JSEntry + 148
    frame #25: 0x000000010097cd60 nsolid`v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [inlined] v8::internal::GeneratedCode<unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, long, unsigned long**>::Call(this=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>) at simulator.h:154:12 [opt]
    frame #26: 0x000000010097cd5c nsolid`v8::internal::(anonymous namespace)::Invoke(isolate=0x0000000110008000, params=0x000000016fdfe1b8)::InvokeParams const&) at execution.cc:427:33 [opt]
    frame #27: 0x000000010097bf48 nsolid`v8::internal::Execution::Call(isolate=0x0000000110008000, callable=Handle<v8::internal::Object> @ x23, receiver=Handle<v8::internal::Object> @ x22, argc=4, argv=0x00006000002291a0) at execution.cc:529:10 [opt]
    frame #28: 0x00000001006db5a8 nsolid`v8::Function::Call(this=<unavailable>, context=<unavailable>, recv=(val_ = 0x0000000110008210), argc=4, argv=0x00006000002291a0) at api.cc:5547:7 [opt]
    frame #29: 0x00000001001a74e8 nsolid`node::builtins::BuiltinLoader::CompileAndCall(this=0x000000010a82fa60, context=(val_ = 0x000000010a87eb00), id="internal/main/run_main_module", argc=4, argv=0x00006000002291a0, optional_realm=0x000000010b00ac60) at node_builtins.cc:482:14
    frame #30: 0x00000001001a736c nsolid`node::builtins::BuiltinLoader::CompileAndCall(this=0x000000010a82fa60, context=(val_ = 0x000000010a87eb00), id="internal/main/run_main_module", realm=0x000000010b00ac60) at node_builtins.cc:466:10
    frame #31: 0x00000001002fd3e0 nsolid`node::Realm::ExecuteBootstrapper(this=0x000000010b00ac60, id="internal/main/run_main_module") at node_realm.cc:161:32
    frame #32: 0x0000000100164604 nsolid`node::StartExecution(env=0x000000010a82f000, main_script_id="internal/main/run_main_module") at node.cc:277:35
    frame #33: 0x000000010016440c nsolid`node::StartExecution(env=0x000000010a82f000, cb=node::StartExecutionCallback @ 0x000000016fdfe938)>) at node.cc:366:12
    frame #34: 0x00000001000156a0 nsolid`node::LoadEnvironment(env=0x000000010a82f000, cb=node::StartExecutionCallback @ 0x000000016fdfea48)>) at environment.cc:549:10
    frame #35: 0x0000000100270304 nsolid`node::NodeMainInstance::Run(this=0x000000016fdfeb80, exit_code=0x000000016fdfead4, env=0x000000010a82f000) at node_main_instance.cc:110:7
    frame #36: 0x000000010026ff50 nsolid`node::NodeMainInstance::Run(this=0x000000016fdfeb80) at node_main_instance.cc:90:3
    frame #37: 0x0000000100166e20 nsolid`node::StartInternal(argc=2, argv=0x00006000030043f0) at node.cc:1397:24
    frame #38: 0x0000000100166a68 nsolid`node::Start(argc=2, argv=0x000000016fdfef98) at node.cc:1404:27
    frame #39: 0x00000001020141c4 nsolid`main(argc=2, argv=0x000000016fdfef98) at node_main.cc:97:10
    frame #40: 0x000000018fbbff28 dyld`start + 2236
  thread #2
    frame #0: 0x000000018fede0a0 libsystem_kernel.dylib`kevent + 8
    frame #1: 0x00000001015dab58 nsolid`uv__io_poll(loop=0x000000010a81b8e8, timeout=-1) at kqueue.c:257:12
    frame #2: 0x00000001015b8338 nsolid`uv_run(loop=0x000000010a81b8e8, mode=UV_RUN_DEFAULT) at core.c:453:5
    frame #3: 0x00000001002d4b88 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Run(this=0x000000010a81b800) at node_platform.cc:95:5
    frame #4: 0x00000001002d4a20 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Start(this=0x000000010a81b800, data=0x000000010a81b800)::'lambda'(void*)::operator()(void*) const at node_platform.cc:64:49
    frame #5: 0x00000001002d49f8 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Start(data=0x000000010a81b800)::'lambda'(void*)::__invoke(void*) at node_platform.cc:63:25
    frame #6: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #3
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000170e0af38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08000) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #4
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000171612f38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08030) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #5
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000171e1af38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08060) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #6
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000172622f38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08090) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #7
    frame #0: 0x000000018fede0a0 libsystem_kernel.dylib`kevent + 8
    frame #1: 0x00000001015dab58 nsolid`uv__io_poll(loop=0x0000000105943bf8, timeout=100) at kqueue.c:257:12
    frame #2: 0x00000001015b8338 nsolid`uv_run(loop=0x0000000105943bf8, mode=UV_RUN_DEFAULT) at core.c:453:5
    frame #3: 0x00000001004f46b8 nsolid`node::nsolid::EnvList::env_list_routine_((null)=0x0000000105944168, envlist=0x0000000105943bd8) at nsolid_api.cc:1490:8
    frame #4: 0x00000001005161ec nsolid`void nsuv::ns_thread::create_proxy_<void (*)(nsuv::ns_thread*, node::nsolid::EnvList*), node::nsolid::EnvList>(arg=0x0000000105944168) at nsuv-inl.h:1829:3
    frame #5: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #8
    frame #0: 0x000000018fed7ed0 libsystem_kernel.dylib`semaphore_wait_trap + 8
    frame #1: 0x00000001015d2744 nsolid`uv_sem_wait(sem=0x0000000105944cec) at thread.c:471:9
    frame #2: 0x00000001005517c0 nsolid`node::inspector::(anonymous namespace)::StartIoThreadMain(unused=0x0000000000000000) at inspector_agent.cc:85:5
    frame #3: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148

```

**Review by juanarbol — COMMENTED** [review]
The next revision addresses all the comments except the "emplace and check" in the thread map for heap snapshots due to a crash.

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 17]
```diff
@@ -12,6 +12,36 @@ NSolidHeapSnapshot::NSolidHeapSnapshot() {
...
+    return UV_EEXIST;
+  }
+  int status = RunCommand(envinst,
+                          CommandType::Interrupt,
+                          start_tracking_heapobjects,
+                          trackAllocations,
+                          duration,
+                          this);
+  // Consider this as taking a heap snapshot in the thread
+  if (status != 0) {
```
Comment (trevnorris) [inline src/nsolid/nsolid_heap_snapshot.cc:37]: Not sure how you come to this conclusion. `RunCommand()` can return `UV_ENOMEM`, which doesn't have anything to do with taking the heap snapshot in this thread.

File: test/addons/nsolid-track-heap-objects/binding.cc [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+static void StopTrackingHeapObjects(const FunctionCallbackInfo<Value>& args) {
+  v8::HandleScope handle_scope(args.GetIsolate());
+  // thread_id
+  assert(args[0]->IsUint32());
+
+  uint64_t thread_id = args[0].As<Number>()->Value();
+
+  int ret = node::nsolid::Snapshot::StopTrackingHeapObjects(
+      node::nsolid::GetEnvInst(thread_id));
+  args.GetReturnValue().Set(Integer::New(args.GetIsolate(), ret));
```
Comment (trevnorris) [inline test/addons/nsolid-track-heap-objects/binding.cc]: `Set()` accepts `int32_t`. Don't need to pass it `Integer::New()`. Just `Set(ret)` should be enough.

**Review by trevnorris — COMMENTED** [review]
Awesome stuff Juan. I just have one question, and one minor adjustment. Linter also fails on the test. Mind taking care of those and I'll approve.

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 17 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc:37]: That's right, I'll delete that comment

File: test/addons/nsolid-track-heap-objects/binding.cc (hunk 18 above)
Comment (juanarbol) [reply test/addons/nsolid-track-heap-objects/binding.cc]: Good to know! I will delete that

**Review by trevnorris — APPROVED** [review]

**Review by santigimeno — APPROVED** [review]
